File: backend/school_service/api/routes/notifications.py
"""API routes for in-app Notifications."""

from fastapi import APIRouter, Depends, Query, status, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional

from school_service.core.database import get_db
from school_service.api.routes.auth import get_current_user
from school_service.services.notification_service import NotificationService
from shared.schemas.user import UserResponse
from shared.schemas.notification import (
    NotificationResponse,
    PaginatedNotificationResponse,
    UnreadCountResponse,
    NotificationUpdate
)

router = APIRouter(prefix="/api/v1/notifications", tags=["notifications"])


@router.get("", response_model=PaginatedNotificationResponse)
async def list_notifications(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    is_read: Optional[bool] = Query(None),
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """List notifications for the current user and school."""
    svc = NotificationService(db)
    return await svc.list_notifications(
        school_id=current_user.school_id,
        user_id=current_user.id,
        is_read=is_read,
        page=page,
        page_size=page_size
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Get the count of unread notifications."""
    svc = NotificationService(db)
    count = await svc.get_unread_count(current_user.school_id, current_user.id)
    return UnreadCountResponse(count=count)


@router.put("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_as_read(
    notification_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Mark a notification as read."""
    svc = NotificationService(db)
    success = await svc.mark_as_read(notification_id, current_user.school_id)
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    return None


@router.post("/mark-all-read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_all_read(
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Mark all notifications for current user as read."""
    svc = NotificationService(db)
    await svc.mark_all_as_read(current_user.school_id, current_user.id)
    return None

File: frontend/lib/api/notifications.ts
/**
 * API client functions for in-app Notifications.
 * 
 * Base URL: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'
 */

import axios from "axios";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface Notification {
    id: number;
    school_id: number;
    user_id: number | null;
    title: string;
    message: string;
    type: "system" | "attendance" | "enrollment";
    link: string | null;
    is_read: boolean;
    created_at: string;
}

export interface PaginatedNotificationResponse {
    items: Notification[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

export interface UnreadCountResponse {
    count: number;
}

/**
 * List notifications for the current user.
 */
export async function listNotifications(
    token: string,
    params: { page?: number; page_size?: number; is_read?: boolean } = {}
): Promise<PaginatedNotificationResponse> {
    const queryParams = new URLSearchParams();
    if (params.page) queryParams.append("page", params.page.toString());
    if (params.page_size) queryParams.append("page_size", params.page_size.toString());
    if (params.is_read !== undefined) queryParams.append("is_read", params.is_read.toString());

    const { data } = await axios.get<PaginatedNotificationResponse>(
        `${API_BASE_URL}/api/v1/notifications?${queryParams.toString()}`,
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        }
    );
    return data;
}

/**
 * Get the count of unread notifications.
 */
export async function getUnreadCount(token: string): Promise<number> {
    console.log(`[NotificationAPI] getUnreadCount calling: ${API_BASE_URL}/api/v1/notifications/unread-count`);
    console.log(`[NotificationAPI] Token present: ${!!token}, length: ${token?.length}, prefix: ${token?.substring(0, 10)}...`);

    const { data } = await axios.get<UnreadCountResponse>(
        `${API_BASE_URL}/api/v1/notifications/unread-count`,
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        }
    );
    return data.count;
}

/**
 * Mark a single notification as read.
 */
export async function markAsRead(token: string, notificationId: number): Promise<void> {
    await axios.put(
        `${API_BASE_URL}/api/v1/notifications/${notificationId}/read`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        }
    );
}

/**
 * Mark all notifications as read.
 */
export async function markAllAsRead(token: string): Promise<void> {
    await axios.post(
        `${API_BASE_URL}/api/v1/notifications/mark-all-read`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        }
    );
}

/** Response from test SMS endpoint (uses school's notification settings). */
export interface TestSmsResponse {
    success: boolean;
    detail: string;
}

/**
 * Send a test message using the current school's notification settings.
 * Optionally specify channel: sms, whatsapp, or both. Defaults to school's parent_delivery.
 */
export async function sendTestSms(
    token: string,
    to: string,
    channel?: "sms" | "whatsapp" | "both"
): Promise<TestSmsResponse> {
    const body: { to: string; channel?: string } = { to };
    if (channel) body.channel = channel;
    const { data } = await axios.post<TestSmsResponse>(
        `${API_BASE_URL}/api/v1/notifications/test-sms`,
        body,
        {
            headers: {
                Authorization: `Bearer ${token}`,
                "Content-Type": "application/json",
            },
        }
    );
    return data;
}

export interface SendWeeklyRemindersResponse {
    sent: number;
    failed: number;
    detail: string;
}

/**
 * Send weekly attendance reminders to all teachers (present X/5 days, percentage).
 */
export async function sendWeeklyReminders(token: string): Promise<SendWeeklyRemindersResponse> {
    const { data } = await axios.post<SendWeeklyRemindersResponse>(
        `${API_BASE_URL}/api/v1/notifications/send-weekly-reminders`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                "Content-Type": "application/json",
            },
        }
    );
    return data;
}
